File: src/backend/cpp/code_sequence.cpp
#include "code_sequence.hpp"

static bool is_repeated(const std::vector<CodeNumber>& numbers, const size_t sub_length) {

    for (size_t i = sub_length; i < numbers.size(); ++i) {
        const auto sub_index = i % sub_length;
        if (numbers.at(i) != numbers.at(sub_index)) {
            return false;
        }
    }

    return true;
}

static XYZ next_angle(const XYZ prev, const XYZ curr, const CodeNumber number) {
    if (number % 2 == 0) {
        return prev;
    } else {
        return other_angle(prev, curr);
    }
}

static bool is_valid(const std::vector<CodeNumber>& numbers, const size_t end) {

    // A sequence of numbers is a valid code sequence iff the code angles wrap around
    // correctly at the beginning and end.

    // Strictly speaking, sequences of length one are allowed if they
    // consist of a single even number (odd ones are invalid).
    // However, all these sequences are trivially empty, so there is no point
    // in considering them.

    auto prev = XYZ::X; // On the last number
    auto curr = XYZ::Y; // On the first number

    for (size_t i = 0; i < end; ++i) {

        // The angle of the next number depends on the current number
        const auto next = next_angle(prev, curr, numbers.at(i));

        // Update prev and curr for the next loop iteration
        // Be careful about the order you do this
        prev = curr;
        curr = next;
    }

    return prev == XYZ::X && curr == XYZ::Y;
}

// Return the end index of the smallest repeated legal subvector
static size_t smallest_index(const std::vector<CodeNumber>& code_numbers) {

    const auto size = code_numbers.size();

    // Find the shortest repeated valid sublist,
    // where we look at the sublists code_numbers[0:i]
    for (size_t i = 2; i < size; ++i) {
        // if i divides the length of the code code_numbers
        if (size % i == 0) {
            if (is_repeated(code_numbers, i) && is_valid(code_numbers, i)) {
                return i;
            }
        }
    }

    // There are no repeaters, so just return the entire vector
    return size;
}

static void minimal_rotation(std::vector<CodeNumber>& code_numbers) {

    // This must be a copy, not a reference, since we mutate code_numbers
    auto min = code_numbers;

    const auto size = code_numbers.size();

    for (size_t i = 0; i < size; ++i) {
        // rotate left
        std::rotate(std::begin(code_numbers), std::next(std::begin(code_numbers)), std::end(code_numbers));

        if (code_numbers < min) {
            min = code_numbers;
        }
    }

    // After size rotations, the vector is now back to where it was before.
    // Now we reverse it, and do it again

    std::reverse(std::begin(code_numbers), std::end(code_numbers));

    for (size_t i = 0; i < size; ++i) {
        // rotate left
        std::rotate(std::begin(code_numbers), std::next(std::begin(code_numbers)), std::end(code_numbers));

        if (code_numbers < min) {
            min = code_numbers;
        }
    }

    code_numbers = min;
}

static void validate(const std::vector<CodeNumber>& numbers) {

    if (numbers.empty()) {
        throw std::runtime_error("CodeSequence: empty code numbers");
    }

    const auto pos = [](const auto num) { return num > 0; };

    const auto all_pos = std::all_of(std::cbegin(numbers), std::cend(numbers), pos);

    if (!all_pos) {
        throw std::runtime_error("CodeSequence: non-positive numbers");
    }

    const auto valid = is_valid(numbers, numbers.size());

    if (!valid) {
        throw std::runtime_error("CodeSequence: invalid parity pattern");
    }
}

CodeSequence::CodeSequence(const std::vector<CodeNumber>& numbers) {

    validate(numbers);

    const auto index = smallest_index(numbers);

    for (size_t i = 0; i < index; ++i) {
        code_numbers.push_back(numbers.at(i));
    }

    minimal_rotation(code_numbers);
}

CodeSequence::const_iterator CodeSequence::begin() const {
    return std::begin(code_numbers);
}

CodeSequence::const_iterator CodeSequence::end() const {
    return std::end(code_numbers);
}

static bool is_palindrome(const std::vector<CodeNumber>& code_numbers, const size_t i) {

    const size_t length = code_numbers.size();
    const size_t half_length = length / 2;

    // i is the index of E1
    // we start at E2, because it wraps around correctly when using %
    // j starts at E2 - 1, and goes <- through l1
    // k starts at E2 + 1, and goes -> and wraps around through l2
    size_t j = i + half_length - 1;
    size_t k = (i + half_length + 1) % length;

    // j and k meet at E1
    while (j != k) {
        if (code_numbers.at(j) != code_numbers.at(k)) {
            return false;
        }

        j -= 1;
        k = (k + 1) % length;
    }

    return true;
}

boost::optional<size_t> CodeSequence::perp_index() const {

    // The sum of an odd sequence is odd, and then doubling it makes it even
    // (with a single factor of 2). Perpendicular sequences always have a sum
    // that is a multiple of 4, so odd sequences are never perpendicular
    // (but what if you double it again?)
    if (is_odd()) {
        return boost::none;
    }

    const auto length = code_numbers.size();
    const auto half_length = length / 2;

    // return the index of E1
    for (size_t i = 0; i < half_length; ++i) {
        const auto first = code_numbers.at(i);
        const auto second = code_numbers.at(i + half_length);

        const auto first_even = (first % 2 == 0);
        const auto second_even = (second % 2 == 0);

        if (first_even && second_even) {
            // Now check if the two lists between the first and second integers are
            // reverses of each other
            // so we could have __ E1 ~~~~~ E2 ---, for example, and then
            // l1 = ~~~~~
            // l2 = ---__
            // need l1 == l2.reverse()

            // check if l1 == l2.reverse()
            if (is_palindrome(code_numbers, i)) {
                return i;
            }
        }
    }

    // Not a closed sequence, so none
    return boost::none;
}

LinComArrZ<XYEta> CodeSequence::constraint(const InitialAngles& initial_angles) const {

    const auto first = initial_angles.first;
    const auto second = initial_angles.second;

    // all odd sequences are stable
    if (is_odd()) {
        // Return zero
        return LinComArrZ<XYEta>{};
    }

    // even means add
    // odd means sub
    LinComArrZ<XYEta> constraint{};
    constraint.add(code_numbers.at(0), xyz_to_xyeta(first));
    constraint.sub(code_numbers.at(1), xyz_to_xyeta(second));

    // We will have to update these as we go along
    auto prev_prev = first;
    auto prev = second;

    for (size_t i = 2; i < code_numbers.size(); ++i) {

        // Which angle this is depends on the previous number
        const auto prev_number = code_numbers.at(i - 1);

        const auto current = next_angle(prev_prev, prev, prev_number);

        const auto current_number = code_numbers.at(i);

        if (i % 2 == 0) {
            // i is even, so add
            constraint.add(current_number, xyz_to_xyeta(current));
        } else {
            // i is odd, so subtract
            constraint.sub(current_number, xyz_to_xyeta(current));
        }

        // Update prev_prev and prev for the next loop iteration
        // Be very careful about the order you do this
        prev_prev = prev;
        prev = current;
    }

    constraint.divide_content();
    constraint.divide_unit();

    return constraint;
}

std::vector<XYZ> CodeSequence::angles(const InitialAngles& initial_angles) const {

    // We start off with these two
    std::vector<XYZ> code_angles{initial_angles.first, initial_angles.second};

    for (size_t i = 2; i < code_numbers.size(); ++i) {

        const auto prev_number = code_numbers.at(i - 1);

        if (prev_number % 2 == 0) {
            // even number
            // second last one
            const auto current = code_angles.at(i - 2);
            code_angles.push_back(current);
        } else {
            // odd number
            const auto prev_prev = code_angles.at(i - 2);
            const auto prev = code_angles.at(i - 1);
            const auto current = other_angle(prev_prev, prev);
            code_angles.push_back(current);
        }
    }

    return code_angles;
}

size_t CodeSequence::length() const {
    return code_numbers.size();
}

CodeNumber CodeSequence::sum() const {

    CodeNumber s = 0;

    for (const auto num : code_numbers) {
        s += num;
    }

    return s;
}

bool CodeSequence::is_stable() const {
    // The angles we choose here are arbitrary
    return constraint(InitialAngles{XYZ::X, XYZ::Y}).is_zero();
}

// The sum and length of the code sequence have the same parity
bool CodeSequence::is_odd() const {
    return code_numbers.size() % 2 != 0;
}

bool CodeSequence::is_perp() const {
    return static_cast<bool>(perp_index());
}

CodeType CodeSequence::type() const {

    const auto stable = is_stable();
    const auto odd = is_odd();
    const auto perp = is_perp();

    if (stable && odd && !perp) {
        return CodeType::SON;
    } else if (stable && !odd && perp) {
        return CodeType::SEP;
    } else if (stable && !odd && !perp) {
        return CodeType::SEN;
    } else if (!stable && !odd && perp) {
        return CodeType::UEP;
    } else if (!stable && !odd && !perp) {
        return CodeType::UEN;
    } else {
        std::ostringstream err{};
        err << "CodeSequence::type: unable to determine type of " << code_numbers;
        throw std::runtime_error(err.str());
    }
}

CodeNumber CodeSequence::number(const size_t i) const {
    return code_numbers.at(i);
}

static int compare(const CodeSequence& lhs, const CodeSequence& rhs) {

    const auto lhs_size = lhs.length();
    const auto rhs_size = rhs.length();

    const auto size_compare = compare(lhs_size, rhs_size);

    if (size_compare != 0) {
        return size_compare;
    } else {
        // Same size, so compare elements one by one
        for (size_t i = 0; i < lhs_size; ++i) {
            const auto lhs_elem = lhs.number(i);
            const auto rhs_elem = rhs.number(i);

            const auto elem_compare = compare(lhs_elem, rhs_elem);

            if (elem_compare != 0) {
                return elem_compare;
            }

            // Else they are equal, so continue to the next element
        }

        // All elements are identical, so they are equal
        return 0;
    }
}

bool operator==(const CodeSequence& lhs, const CodeSequence& rhs) {
    return compare(lhs, rhs) == 0;
}

bool operator<(const CodeSequence& lhs, const CodeSequence& rhs) {
    return compare(lhs, rhs) < 0;
}

std::ostream& operator<<(std::ostream& os, const CodeSequence& code_seq) {

    bool first = true;

    for (const auto num : code_seq) {

        if (!first) {
            os << ' ';
        }

        os << num;

        first = false;
    }

    return os;
}

File: src/java/billiards/codeseq/CodeSequence.java
package billiards.codeseq;

import billiards.math.LinCom;
import billiards.math.XYEta;
import billiards.math.XYZ;

import org.eclipse.collections.api.list.primitive.ImmutableIntList;
import org.eclipse.collections.api.list.primitive.IntList;
import org.eclipse.collections.api.list.primitive.MutableIntList;
import org.eclipse.collections.impl.list.mutable.primitive.IntArrayList;

import java.util.Optional;

// In Java, the type system isn't expressive enough to describe immutable
// or constant objects, so you deal with this by making immutability
// a feature of the API of the class. Aka, private members, and no
// methods that return a mutable reference to or mutate those members.
// As such, you have the common pattern of a mutable "builder" object,
// which you construct, and then you create an immutable object from
// the builder. In Rust and C++, this is unnecessary.

// You can't modify an IntList from the binding, but it could
// be modified by performing some other operation.
// But you can't modify an ImmutableIntList from anywhere.
// That is the difference. So use the first for function parameters,
// and the second (or MutableIntList) for variables.
// Use the interface when specifying the type, the specific
// class when creating the type

// We us an ImmutableIntList. It gives us a fixed length, immutable list.
// It doesn't have nice indexing, but that's stupid 'ol java for you

public final class CodeSequence implements Comparable<CodeSequence> {

    public final ImmutableIntList codeNumbers;

    public CodeSequence(final IntList tmpCodeNumbers) {

        validate(tmpCodeNumbers);

        final int index = smallestIndex(tmpCodeNumbers);

        final MutableIntList sublist = subList(tmpCodeNumbers, 0, index);

        this.codeNumbers = minimalRotation(sublist);
    }

    private static MutableIntList subList(final IntList list, final int start, final int end) {

        final MutableIntList subList = new IntArrayList();
        for (int i = start; i < end; ++i) {
            final int elem = list.get(i);
            subList.add(elem);
        }

        return subList;
    }

    private static boolean isRepeated(final IntList codeNumbers, final int subLength) {

        for (int i = subLength; i < codeNumbers.size(); ++i) {
            final int subIndex = i % subLength;
            if (codeNumbers.get(i) != codeNumbers.get(subIndex)) {
                return false;
            }
        }

        return true;
    }

    private static void validate(final IntList codeNumbers) {

        // Must be nonempty
        if (codeNumbers.isEmpty()) {
            throw new RuntimeException("empty code numbers");
        }

        // All numbers must be strictly positive
        final boolean allPos = codeNumbers.allSatisfy(num -> num > 0);
        if (!allPos) {
            throw new RuntimeException("non-positive numbers");
        }

        // Check if it has a legal pattern
        final boolean legal = isLegal(codeNumbers, codeNumbers.size());

        if (!legal) {
            throw new RuntimeException("illegal pattern");
        }
    }

    // codeNumbers must already be valid
    private static int smallestIndex(final IntList codeNumbers) {

        final int size = codeNumbers.size();

        // Find the shortest repeated legal sublist.
        for (int i = 2; i < size; ++i) {
            if (size % i == 0) {
                if (isRepeated(codeNumbers, i) && isLegal(codeNumbers, i)) {
                    return i;
                }
            }
        }

        // Otherwise, there are no repeaters, so just return the code numbers
        return size;
    }

    private static XYZ nextAngle(final XYZ prev, final XYZ curr, final int number) {
        if (number % 2 == 0) {
            return prev;
        } else {
            return XYZ.otherAngle(prev, curr);
        }
    }

    private static boolean isLegal(final IntList codeNumbers, final int end) {

        XYZ prev = XYZ.X;
        XYZ curr = XYZ.Y;

        for (int i = 0; i < end; ++i) {
            final int num = codeNumbers.get(i);

            final XYZ next = nextAngle(prev, curr, num);

            prev = curr;
            curr = next;
        }

        return prev == XYZ.X && curr == XYZ.Y;
    }

    private static ImmutableIntList minimalRotation(final MutableIntList codeNumbers) {

        // We rotate and reverse `codeNumbers`, which would change `min` too if it were
        // just a reference. To prevent this, `min` is an immutable copy
        ImmutableIntList min = codeNumbers.toImmutable();

        final int length = codeNumbers.size();

        // Rotate length times
        for (int i = 0; i < length; ++i) {
            rotateLeft(codeNumbers);

            // if codeNumbers < min
            if (compareIntList(codeNumbers, min) < 0) {
                min = codeNumbers.toImmutable();
            }
        }

        // After length rotations, the list is now back
        // to where it was before. Rotate it, and do it again

        codeNumbers.reverseThis();

        for (int i = 0; i < length; ++i) {
            rotateLeft(codeNumbers);

            // if codeNumbers < min
            if (compareIntList(codeNumbers, min) < 0) {
                min = codeNumbers.toImmutable();
            }
        }

        return min;
    }

    private static int compareIntList(final IntList list1, final IntList list2) {
        final int lengthComp = Integer.compare(list1.size(), list2.size());

        if (lengthComp < 0) {
            // list1.size() < list2.size()
            return -1;
        } else if (lengthComp > 0) {
            // list1.size() > list2.size()
            return 1;
        } else {
            // sizes are the same

            for (int i = 0; i < list1.size(); ++i) {
                final int elem1 = list1.get(i);
                final int elem2 = list2.get(i);

                final int comp = Integer.compare(elem1, elem2);

                if (comp < 0) {
                    return -1;
                } else if (comp > 0) {
                    return 1;
                }

                // else comp == 0, so check the next element
            }

            return 0;
        }
    }

    // Rotate a list left by one spot
    private static void rotateLeft(final MutableIntList list) {

        if (list.notEmpty()) {
            final int first = list.get(0);
            for (int i = 1; i < list.size(); ++i) {
                final int elem = list.get(i);
                list.set(i - 1, elem);
            }

            list.set(list.size() - 1, first);
        }
    }

    public int sum() {

        int sum = 0;
        for (int i = 0; i < this.codeNumbers.size(); ++i) {
            final int codeNumber = this.codeNumbers.get(i);
            sum = Math.addExact(sum, codeNumber);
        }

        return sum;
    }

    public int length() {
        return this.codeNumbers.size();
    }

    public CodeType type() {
        final boolean odd = isOdd();
        final boolean closed = isClosed();
        final boolean stable = isStable();

        if (!closed && stable && odd) {
            return CodeType.OSO;
        } else if (!closed && stable && !odd) {
            return CodeType.OSNO;
        } else if (!closed && !stable) {
            return CodeType.ONS;
        } else if (closed && stable) {
            return CodeType.CS;
        } else if (closed && !stable) {
            return CodeType.CNS;
        } else {
            throw new RuntimeException(codeNumbers + " cannot be classified");
        }
    }

    public Optional<Integer> closedIndex() {

        // Odd code sequences are never closed
        if (isOdd()) {
            return Optional.empty();
        }

        final int length = this.codeNumbers.size();
        final int halfLength = length / 2;

        // Iterate over codeNumbers in intervals of halfLength, checking
        // to see if we find two even numbers.
        for (int i = 0; i < halfLength; i += 1) {
            final int first = this.codeNumbers.get(i);
            final int second = this.codeNumbers.get(i + halfLength);

            final boolean firstEven = first % 2 == 0;
            final boolean secondEven = second % 2 == 0;

            if (firstEven && secondEven) {
                // Now check if the two lists between the first and second integers are
                // reverses of each other.
                // We could have __ E ~~~~~ E ---, for example, and then
                // a = ~~~~~
                // b = ---__ (and then we reverse it)
                if (isPalindrome(this.codeNumbers, i)) {
                    return Optional.of(i);
                }
            }
        }

        return Optional.empty();
    }

    private static boolean isPalindrome(final IntList codeNumbers, final int i) {

        final int length = codeNumbers.size();
        final int halfLength = length / 2;

        // i is the index of E1
        // we start at E2, because it wraps around correctly when using %
        // j starts at E2 - 1, and goes <- through l1
        // k starts at E2 + 1, and goes -> and wraps around through l2
        int j = i + halfLength - 1;
        int k = (i + halfLength + 1) % length;

        // j and k meet at E1
        while (j != k) {
            if (codeNumbers.get(j) != codeNumbers.get(k)) {
                return false;
            }

            j -= 1;
            k = (k + 1) % length;
        }

        return true;
    }

    public boolean isClosed() {
        return closedIndex().isPresent();
    }

    public boolean isOdd() {
        return this.codeNumbers.size() % 2 != 0;
    }

    public boolean isStable() {
        // The angles we use here are arbitrary
        return constraint(XYZ.X, XYZ.Y).isZero();
    }

    public LinCom<XYEta> constraint(final XYZ first, final XYZ second) {

        if (isOdd()) {
            // odd codes are stable, so 0
            return new LinCom<>(0, 0, 0);
        }

        // even indices are add
        // odd indices are sub

        final LinCom<XYZ> constraint = new LinCom<>(0, 0, 0);
        constraint.add(codeNumbers.get(0), first);
        constraint.sub(codeNumbers.get(1), second);

        XYZ prevPrev = first;
        XYZ prev = second;

        for (int i = 2; i < codeNumbers.size(); ++i) {
            final int prevNumber = codeNumbers.get(i - 1);

            final XYZ current = nextAngle(prevPrev, prev, prevNumber);

            final int currentNumber = codeNumbers.get(i);

            if (i % 2 == 0) {
                // even index, so add
                constraint.add(currentNumber, current);
            } else {
                // odd index, so sub
                constraint.sub(currentNumber, current);
            }

            // Update prev_prev and prev for the next loop iteration
            prevPrev = prev;
            prev = current;
        }

        // z = 2*eta - x - y

        final int xCoeff = constraint.coeff(XYZ.X) - constraint.coeff(XYZ.Z);
        final int yCoeff = constraint.coeff(XYZ.Y) - constraint.coeff(XYZ.Z);
        final int etaCoeff = 2 * constraint.coeff(XYZ.Z);

        final LinCom<XYEta> etaConstraint = new LinCom<>(xCoeff, yCoeff, etaCoeff);
        etaConstraint.divideContent();
        etaConstraint.normalizeUnit();

        return etaConstraint;
    }

    @Override
    public String toString() {
        return this.codeNumbers.makeString(" ");
    }

    @Override
    public int hashCode() {
        return this.codeNumbers.hashCode();
    }

    @Override
    public boolean equals(final Object obj) {
        final CodeSequence other = (CodeSequence) obj;
        return this.codeNumbers.equals(other.codeNumbers);
    }

    @Override
    public int compareTo(final CodeSequence other) {
        return compareIntList(this.codeNumbers, other.codeNumbers);
    }
}
